src/utils/WebSocketsService.ts

```diff
@@ -69,3 +74,106 @@ export default {
   isConnected,
   send,
 };
+
+/**
+ * This hook exposes SocketWrapper containing an instance of the created WebSocket object
+ * and the send function. `send` function "wraps" WebSocket's `send` to add
+ * checks on the socket object's ready state before performing actual send.
+ *
+ * @param props { messageHandler, userAuthToken }
+ */
+export function useWebSocket(props) {
+  const { messageHandler, userAuthToken } = props || {};
+
+  /**
+   * Main instance of websocket which events and states are tracked.
+   * This is stored in the instance property of the SocketWrapper - the object that
+   * is passed as a value to the SocketContext defined in the App.tsx
+   */
+  const [instance, setInstance] = useState<WebSocket>(null);
+
+  const [readyState, setReadyState] = useState(undefined);
+
+  const sendFn = (i: WebSocket, token: string) => {
+    return message => {
+      if (i) {
+        i.send(JSON.stringify({ ...message, token }));
+      }
+    };
+  };
+
+  /**
+   * socketWrapper state stores the current instance and the `send` function.
+   * `send` function is run the MiddleWare executes `sendMessage()` defined in
+   * a WebSocket-supported action creator (ie, boardActions#openedBoard)
+   */
+  const [socketWrapper, setSocketWrapper] = useState<ISocketWrapper>({
+    instance,
+    send: sendFn(instance, userAuthToken),
```

Comment:
Yea I actually tried doing that at https://github.com/srcclr/sapling-frontend/pull/32/files/e059422620544762f4e8a7c6514d2e53061478fe#diff-88c874b315e1af93323472bd7718c996R164 and set it initially to `null`. But since the MiddleWare would `send`, even before the instance exists https://github.com/srcclr/sapling-frontend/pull/32/files/e059422620544762f4e8a7c6514d2e53061478fe#diff-88c874b315e1af93323472bd7718c996R163 , the MiddleWare will complain that `send` is null. hmm but I think if I set this initial value to a dummy function instead of null and place the actual function inline, it might be good too 🤔

Reply:
Yeah, that's also possible. So websocket operations are a no-op until the token is present.